### ISSUE TITLE
Frequent CPU spikes lead to battery drain

### ISSUE THREAD

**Issue by Valentin-N** [SimpleMobileTools/Simple-Calendar#1110]
As mentioned in https://github.com/SimpleMobileTools/Simple-Calendar/issues/861, there seems to be something that triggers Simple Calendar (and Calendar Storage) to both spike CPU usage simultaneously and continuously (every few seconds). In my setup I have 2 accounts enabled (Google and Exchange). The CPU spike can start occurring almost immediately after installing the app and enabling CalDAV, or randomly within an hour of installing the app.

Logcat shows the following trace simultaneously with the CPU spikes:

```
04-06 23:28:11.391 E/DatabaseUtils(12831): Writing exception to parcel
04-06 23:28:11.391 E/DatabaseUtils(12831): java.lang.IllegalArgumentException: Color type: 0 and index -1 does not exist for account.
04-06 23:28:11.391 E/DatabaseUtils(12831):      at com.android.providers.calendar.CalendarProvider2.verifyColorExists(CalendarProvider2.java:4488)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at com.android.providers.calendar.CalendarProvider2.updateInTransactionInner(CalendarProvider2.java:4285)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at com.android.providers.calendar.CalendarProvider2.updateInTransaction(CalendarProvider2.java:4185)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at com.android.providers.calendar.SQLiteContentProvider.update(SQLiteContentProvider.java:155)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at com.android.providers.calendar.CalendarProvider2.update(CalendarProvider2.java:2299)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at android.content.ContentProvider$Transport.update(ContentProvider.java:420)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at android.content.ContentProviderNative.onTransact(ContentProviderNative.java:215)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at android.os.Binder.execTransactInternal(Binder.java:1021)
04-06 23:28:11.391 E/DatabaseUtils(12831):      at android.os.Binder.execTransact(Binder.java:994)
```

![IMG_2020-04-07-000412](https://user-images.githubusercontent.com/1926716/78613384-439de600-785b-11ea-93b8-0fced419679b.png)

**Comment by tibbi** [comment]
there isnt even Simple Calendar mentiond at the logcat. What service are you CalDAV syncing through? I mean Google, Nextcloud, or?

**Comment by Valentin-N** [comment]
The traces in the logcat appear a short while after I install Simple Calendar and enable CalDAV. They stop as soon as I force stop Simple Calendar (or uninstall it). I am not saying that is a root cause, but it is a symptom I can reliably observe happening at the same time as Simple Calendar (and Calendar Storage) showing up with increased CPU usage.

I use Google Apps for the Google account and [Nine](https://play.google.com/store/apps/details?id=com.ninefolders.hd3) for the Exchange account. I also have 2 accounts synced via [DAVx⁵](https://f-droid.org/en/packages/at.bitfire.davdroid/) that are disabled and have empty calendars.

**Comment by tibbi** [comment]
Well it seems like some CalDAV sync is triggered too often in the background. Does disabling "CalDAV sync" in the Simple Calendars app settings fix it? I mean instead of uninstalling it.

**Comment by michaelof** [comment]
@Valentin-N how do you establish the lionk between your calDAV based calendars and your phone? DAVx5? 
Because afaik all sync is done between your calDAV connector and Android's Calendar Storage, outside and independent of Simple Calendar.
FYI I'm using Simple Calendar, Simple Contacts, OpenTasks along with DAVx5. Server is Nextcloud, own VPS. On several devices, Android 4.1 (stone-age, older versions but still working), 4.4, 6.0 and 9. Working well, no battery drain.

When I look at my tasks via ssh localhost via connectbot and cmd "top", I see e.g. that my softphone "linphone" and my (bloody cool) KDE-Connect app are triggering a 2-5% cpu spike every couple of seconds. But in power mgmt they're not even taken into account, compared to "Display" etc. when battery is low, let say 10% left. So in my case 90% "used" means ~85% counted and measured, ~5% undefined loss, who cares. 

So a) I'm not sure if your spike is really a "draining issue" and b) my Simple Calender, although having set battery optimization to OFF, because otherwise notifications are not triggered always, Simple Calendar does NOT produce and remarkable cpu spikes.

**Comment by michaelof** [comment]
FYI my DAVx5 sync  settings are 
contacts every 60 min + at once for local changes
events every 30 min + at once for local changes
tasks every 240 min + at once for local changes
sync on wlan and mobile
no history restriction (full sync)

2 adressbooks
5 calendars

**Comment by Valentin-N** [comment]
> Well it seems like some CalDAV sync is triggered too often in the background. Does disabling "CalDAV sync" in the Simple Calendars app settings fix it? I mean instead of uninstalling it.

As soon as I disable CalDAV sync, the CPU usage stops. Once I re-enable it, there's a higher than usual initial CPU usage (I suspect there's some initial sync work going on?), then it stabilizes to the pattern described previously where there's a CPU spike once every few seconds and keeps doing that continuously while Simple Calendar is installed and running (i.e. not force stoped).


> @Valentin-N how do you establish the lionk between your calDAV based calendars and your phone? DAVx5?

As mentioned in my previous post, Google Apps (3 accounts, 1 active), Nine (1 account, active) and DAVx5 (2 accounts, both with sync disabled).

> When I look at my tasks via ssh localhost via connectbot and cmd "top", I see e.g. that my softphone "linphone" and my (bloody cool) KDE-Connect app are triggering a 2-5% cpu spike every couple of seconds. But in power mgmt they're not even taken into account, compared to "Display" etc. when battery is low, let say 10% left. So in my case 90% "used" means ~85% counted and measured, ~5% undefined loss, who cares.
> 
> So a) I'm not sure if your spike is really a "draining issue" and b) my Simple Calender, although having set battery optimization to OFF, because otherwise notifications are not triggered always, Simple Calendar does NOT produce and remarkable cpu spikes.

Using a combination of apps I can confidently say that aside from the obvious CPU spikes there's a consistent increase in real-time reported battery discharge rate while Simple Calendar is running (as measured with [AccuBattery](https://play.google.com/store/apps/details?id=com.digibites.accubattery)). I am seeing an average discharge rate of around 200 mA while the phone is on and idle and Simple Calendar not running. With Simple Calendar running and the phone on and idle I am seeing an average discharge rate of around 450 mA. After leaving Simple Calendar running for a few hours, then uninstalling it, the battery discharge rate curve as reported by Android also changes slope: it is visibly more abrupt while Simple Calendar has been running and becomes more horizontal when Simple Calendar is uninstalled. The phone also becomes noticeably warm to the touch when Simple Calendar is running. I hope this settles the matter of what is a reasonable discharge rate or reasonable background CPU usage, so please don't move the conversation in that direction any further.

For reference, I have 2 other calendar apps installed and in use, none of which cause _any_ noticeable CPU usage (neither for the app, nor Calendar Storage): the built-in AOSP Calendar app and [Etar](https://f-droid.org/en/packages/ws.xsoh.etar/).

**Comment by hbakken** [comment]
I have seen similar problems with my Simple Calendar app.  I think I started it by disabling the Google Calendar app earlier, which probably means I had nothing doing the CalDAV sync (I have Samsung's calendar, although it doesn't appear to do much).  After this (this was some time early in March), I could see my events for March and from June onwards, but nothing from April and May.  Also, Simple Calendar would really spike in CPU usage, and often crash.
I have since installed Google Calendar, which works fine, but Simple Calendar is still as broken.

**Comment by tibbi** [comment]
Any of you are using any widget of Simple Calendar Pro?

**Comment by Valentin-N** [comment]
I use the daily calendar widget

**Comment by tibbi** [comment]
which do you mean? Simple Calendar has no daily widget. Just a monthly and a list one.

**Comment by Valentin-N** [comment]
The list widget (is what I call daily), that's the one I use

**Comment by hbakken** [comment]
I use the pro version.

**Comment by hbakken** [comment]
I also find I can no longer swipe to switch months (in monthly view), I have to use the arrows.  I've tried deleting all data for simple calendar, and to reinstall it, but to no avail.  I haven't rooted my phone, so I don't really have very good debugging options, I'm afraid.

**Comment by tibbi** [comment]
I guess you are just triggering the Pull to refresh, try swiping a bit upwards too. Anyway, dont go offtopic here.

**Comment by hbakken** [comment]
I think it might be related.  It feels like it sort-of tries, but is bogged down.  Anyway, it could be a totally unrelated problem, of course.

**Comment by hbakken** [comment]
I've now found that Simple Calendar shows the same problems on my tablet, where I haven't done anything to Google Calendar, so it seems unrelated.  I guess it could be related to some of my events, somehow, but hard to tell which...

**Comment by tibbi** [comment]
are you using any widgets on the devices where the spikes occur?

**Comment by hbakken** [comment]
What kind of widgets?  Anyway, I don't really have spikes as such, it is constant for me whenever Simple Calendar runs, it's dog slow, does not sync April and May this year (but does seem to sync later months).
I guess it could be related to some of the events I have in April or May (or both).

**Comment by tibbi** [comment]
Well home screen widgets like the monthly view or the event list view

**Comment by hbakken** [comment]
No, I don't have any of those.  Just the shortcut or whatever it's called to Simple Calendar itself.

**Comment by sophiekuehn** [comment]
Have to report a similar if not the same issue.  Newest Simple Calendar Pro from F-Droid. Android 8.0. One DAVx5 sync and two google accounts. Frequent battery hog alarm. 

![Screenshot_20201231-165353_Device maintenance](https://user-images.githubusercontent.com/6147076/103417014-4ed33780-4b89-11eb-853f-ee4bb9d27074.jpg)

**Comment by tibbi** [comment]
should be fixed in 6.13.0
